### PR TITLE
Upgrade @itwin/eslint-plugin to 4.0.0-dev.43 and eslint to 8.x accordingly

### PR DESCRIPTION
**Pull request by anmolshres98**
Change motivated from Nightman build failures owing to eslint-plugin version mismatch: https://github.com/iTwin/nightman/actions/runs/5514259302/jobs/10053301272

Specifically, newer eslint-plugin uses a [new config system](https://eslint.org/blog/2022/08/new-config-system-part-2/) and our packages need to be updated to use this new system

### CODE REVIEW

File: apps/load-tests/backend/eslint.config.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+      jsx: true,
+      modules: true,
+    },
+  },
+};
+
+module.exports = [
+  ...eslintBaseConfig,
+  {
+    files: ["**/*.{ts,tsx}"],
```
Comment (grigasp) [inline apps/load-tests/backend/eslint.config.js:20]: For backend configs we don't need JSX
```suggestion
      modules: true,
    },
  },
};

module.exports = [
  ...eslintBaseConfig,
  {
    files: ["**/*.ts"],
```

Comment (grigasp) [reply apps/load-tests/backend/eslint.config.js:20]: This also applies to a number of other configs for packages that don't use React.

File: apps/test-app/frontend/eslint.config.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+const iTwinPlugin = require("@itwin/eslint-plugin");
+const eslintBaseConfig = require("../../../eslint.base.config");
+const jsxRuntimeConfig = require("eslint-plugin-react");
```
Comment (grigasp) [inline apps/test-app/frontend/eslint.config.js:3]: `package.json` doesn't specify `eslint-plugin-react` as a dependency

File: packages/components/eslint.config.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
+const iTwinPlugin = require("@itwin/eslint-plugin");
+const eslintBaseConfig = require("../../eslint.base.config");
+const jsxRuntimeConfig = require("eslint-plugin-react");
+
+
```
Comment (grigasp) [inline packages/components/eslint.config.js:5]: extra new line

File: packages/components/eslint.config.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
+const iTwinPlugin = require("@itwin/eslint-plugin");
+const eslintBaseConfig = require("../../eslint.base.config");
+const jsxRuntimeConfig = require("eslint-plugin-react");
```
Comment (grigasp) [inline packages/components/eslint.config.js:3]: `package.json` doesn't have `eslint-plugin-react` as a dev dependency

File: packages/opentelemetry/eslint.config.js [hunk 5]
```diff
@@ -0,0 +1,8 @@
+const iTwinPlugin = require("@itwin/eslint-plugin");
+
+module.exports = [
```
Comment (grigasp) [inline packages/opentelemetry/eslint.config.js:3]: I believe this config should be similar to other non-ui package configs (e.g. the `testing` one) and extend the base config. I see it wasn't doing that, but it should.

File: packages/components/src/presentation-components/instance-filter-builder/InstanceFilterBuilder.tsx [hunk 6]
```diff
@@ -107,18 +107,18 @@ export function usePresentationInstanceFilteringProps(
...
-    | "properties"
-    | "classes"
-    | "selectedClasses"
-    | "onClassSelected"
-    | "onClassDeselected"
-    | "onClearClasses"
-    | "propertyRenderer"
-    | "onRulePropertySelected"
-    | "isDisabled"
+  InstanceFilterBuilderProps,
```
Comment (grigasp) [inline packages/components/src/presentation-components/instance-filter-builder/InstanceFilterBuilder.tsx:110]: Do we know why indent got changed here? I like the previous style better.

**Review by grigasp — CHANGES_REQUESTED** [review]
I'm not sure I'm a fan of single-line member delimiter being a comma rather than semi-colon which we have currently. @saskliutas what do you think?

**Comment by saskliutas** [comment]
> I'm not sure I'm a fan of single-line member delimiter being a comma rather than semi-colon which we have currently. @saskliutas what do you think?

I don't have strong preference on this but it looks like prettier uses ';'. It would be interesting to know why these formatting changes are made after upgrading eslint-plugin package. Is there some changes to rules in `itwinjs-recommended` config? Or previous configuration was not working correctly?

**Comment by grigasp** [comment]
> I don't have strong preference on this but it looks like prettier uses ';'. It would be interesting to know why these formatting changes are made after upgrading eslint-plugin package. Is there some changes to rules in itwinjs-recommended config? Or previous configuration was not working correctly?

I assumed `eslint-plugin@dev.38` had a different config, but I just checked and the `@typescript-eslint/member-delimiter-style` rule config is the same in both `dev.38` and `dev.43`, so I'm not sure what caused that anymore.

**Review by saskliutas — CHANGES_REQUESTED** [review]

File: apps/load-tests/backend/eslint.config.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
+const iTwinPlugin = require("@itwin/eslint-plugin");
```
[the remark this answers is not in the document]
Comment (anmolshres98) [reply apps/load-tests/backend/eslint.config.js:5]: added, good catch. thanks for suggestion

File: apps/load-tests/backend/eslint.config.js (hunk 1 above)
Comment (anmolshres98) [reply apps/load-tests/backend/eslint.config.js]: removed for all relevant packages

File: apps/test-app/frontend/eslint.config.js (hunk 2 above)
Comment (anmolshres98) [reply apps/test-app/frontend/eslint.config.js]: added

File: packages/components/eslint.config.js (hunk 3 above)
Comment (anmolshres98) [reply packages/components/eslint.config.js]: removed

File: packages/components/eslint.config.js (hunk 4 above)
Comment (anmolshres98) [reply packages/components/eslint.config.js]: added

File: packages/opentelemetry/eslint.config.js (hunk 5 above)
Comment (anmolshres98) [reply packages/opentelemetry/eslint.config.js:8]: done

File: packages/components/src/presentation-components/instance-filter-builder/InstanceFilterBuilder.tsx (hunk 6 above)
Comment (anmolshres98) [reply packages/components/src/presentation-components/instance-filter-builder/InstanceFilterBuilder.tsx]: It is from `@typescript-eslint` that is being used by itwin/eslint-plugin

**Comment by anmolshres98** [comment]
> > I don't have strong preference on this but it looks like prettier uses ';'. It would be interesting to know why these formatting changes are made after upgrading eslint-plugin package. Is there some changes to rules in itwinjs-recommended config? Or previous configuration was not working correctly?
> 
> I assumed `eslint-plugin@dev.38` had a different config, but I just checked and the `@typescript-eslint/member-delimiter-style` rule config is the same in both `dev.38` and `dev.43`, so I'm not sure what caused that anymore.

I am not completely certain about this either but you are right. It doesn't seem like `@typescript-eslint` dep changed in itwin/eslint-plugin between the two versions. My guess is either the linting wasn't working properly on this repo properly or something was messing up in itwin/eslint-plugin side. @Jake-Screen is out right now but he would be able to answer this better.

**Comment by anmolshres98** [comment]
> > > I don't have strong preference on this but it looks like prettier uses ';'. It would be interesting to know why these formatting changes are made after upgrading eslint-plugin package. Is there some changes to rules in itwinjs-recommended config? Or previous configuration was not working correctly?
> > 
> > 
> > I assumed `eslint-plugin@dev.38` had a different config, but I just checked and the `@typescript-eslint/member-delimiter-style` rule config is the same in both `dev.38` and `dev.43`, so I'm not sure what caused that anymore.
> 
> I am not completely certain about this either but you are right. It doesn't seem like `@typescript-eslint` dep changed in itwin/eslint-plugin between the two versions. My guess is either the linting wasn't working properly on this repo properly or something was messing up in itwin/eslint-plugin side. @Jake-Screen is out right now but he would be able to answer this better.
> 
> There might be mismatch between prettier rules and eslint rules on this repo. That needs to be configured so that they match and don't conflict

After some investigation I have determined that this is because the rules from `eslint-config-prettier` are not being imported properly. Currently investigating this

**Comment by anmolshres98** [comment]
> > > > I don't have strong preference on this but it looks like prettier uses ';'. It would be interesting to know why these formatting changes are made after upgrading eslint-plugin package. Is there some changes to rules in itwinjs-recommended config? Or previous configuration was not working correctly?
> > > 
> > > 
> > > I assumed `eslint-plugin@dev.38` had a different config, but I just checked and the `@typescript-eslint/member-delimiter-style` rule config is the same in both `dev.38` and `dev.43`, so I'm not sure what caused that anymore.
> > 
> > 
> > I am not completely certain about this either but you are right. It doesn't seem like `@typescript-eslint` dep changed in itwin/eslint-plugin between the two versions. My guess is either the linting wasn't working properly on this repo properly or something was messing up in itwin/eslint-plugin side. @Jake-Screen is out right now but he would be able to answer this better.
> > There might be mismatch between prettier rules and eslint rules on this repo. That needs to be configured so that they match and don't conflict
> 
> After some investigation I have determined that this is because the rules from `eslint-config-prettier` are not being imported properly. Currently investigating this

Update: figured out the fix, pushed, prettier should be working properly with eslint now. The only issue remaining is this one: 
image

I have set rule to warn for now:
image


Let me know how you want to proceed, thanks!

File: packages/components/src/presentation-components/instance-filter-builder/InstanceFilterBuilder.tsx (hunk 6 above)
Comment (anmolshres98) [reply packages/components/src/presentation-components/instance-filter-builder/InstanceFilterBuilder.tsx]: This is from the prettier issue I have explained below, solved now with no change needed

**Comment by grigasp** [comment]
> Update: figured out the fix, pushed, prettier should be working properly with eslint now. The only issue remaining is this one: image
> 
> I have set rule to warn for now: image
> 
> Let me know how you want to proceed, thanks!

I think that's a valid error, went ahead and fixed all occurrences of it.

**Review by grigasp — APPROVED** [review]

**Pull request merged**